test/extensions/filters/http/ext_proc/filter_test.cc

```diff
@@ -1207,6 +1278,265 @@ TEST_F(HttpFilterTest, PostFastAndBigRequestPartialBuffering) {
+  for (uint32_t i = 0; i < chunk_number; i++) {
+    Buffer::OwnedImpl req_data("foo");
+    EXPECT_EQ(FilterDataStatus::Continue, filter_->decodeData(req_data, false));
+    processRequestBody(absl::nullopt, false);
+  }
+
+  EXPECT_EQ(FilterTrailersStatus::StopIteration, filter_->decodeTrailers(request_trailers_));
+  processRequestTrailers(absl::nullopt, true);
+
+  response_headers_.addCopy(LowerCaseString(":status"), "200");
+  EXPECT_EQ(FilterHeadersStatus::StopIteration, filter_->encodeHeaders(response_headers_, false));
+  processResponseHeaders(true, absl::nullopt);
+
+  for (uint32_t i = 0; i < chunk_number * 2; i++) {
+    Buffer::OwnedImpl resp_data("bar");
+    EXPECT_EQ(FilterDataStatus::Continue, filter_->encodeData(resp_data, false));
+    processResponseBody(absl::nullopt, false);
+  }
+
+  EXPECT_EQ(FilterTrailersStatus::StopIteration, filter_->encodeTrailers(response_trailers_));
+  processResponseTrailers(absl::nullopt, false);
+  filter_->onDestroy();
+  EXPECT_EQ(1, config_->stats().streams_started_.value());
+  // Total gRPC messages include two headers and two trailers on top of the req/resp chunk data.
+  uint32_t total_msg = 3 * chunk_number + 4;
+  EXPECT_EQ(total_msg, config_->stats().stream_msgs_sent_.value());
+  EXPECT_EQ(total_msg, config_->stats().stream_msgs_received_.value());
+  EXPECT_EQ(1, config_->stats().streams_closed_.value());
+
+  // Check inbound gRPC call stats.
+  auto& grpc_calls_in = getGrpcCalls(envoy::config::core::v3::TrafficDirection::INBOUND);
+  EXPECT_TRUE(grpc_calls_in.header_stats_ != nullptr);
+  checkGrpcCall(*grpc_calls_in.header_stats_, std::chrono::microseconds(10), Grpc::Status::Ok);
+
+  EXPECT_TRUE(grpc_calls_in.trailer_stats_ != nullptr);
+  checkGrpcCall(*grpc_calls_in.trailer_stats_, std::chrono::microseconds(10), Grpc::Status::Ok);
+
+  EXPECT_TRUE(grpc_calls_in.body_stats_ != nullptr);
+  checkGrpcCallBody(*grpc_calls_in.body_stats_, chunk_number, Grpc::Status::Ok,
+                    std::chrono::microseconds(10) * chunk_number, std::chrono::microseconds(10),
+                    std::chrono::microseconds(10));
+
+  // Check outbound gRPC call stats.
+  auto& grpc_calls_out = getGrpcCalls(envoy::config::core::v3::TrafficDirection::OUTBOUND);
+  EXPECT_TRUE(grpc_calls_out.header_stats_ != nullptr);
+  checkGrpcCall(*grpc_calls_out.header_stats_, std::chrono::microseconds(10), Grpc::Status::Ok);
+
+  EXPECT_TRUE(grpc_calls_out.trailer_stats_ != nullptr);
+  checkGrpcCall(*grpc_calls_out.trailer_stats_, std::chrono::microseconds(10), Grpc::Status::Ok);
+
+  EXPECT_TRUE(grpc_calls_out.body_stats_ != nullptr);
+  checkGrpcCallBody(*grpc_calls_out.body_stats_, 2 * chunk_number, Grpc::Status::Ok,
+                    std::chrono::microseconds(10) * 2 * chunk_number, std::chrono::microseconds(10),
+                    std::chrono::microseconds(10));
+}
+
+// gRPC call fails when streaming sends small chunk request data.
+TEST_F(HttpFilterTest, StreamingSendRequestDataGrpcFail) {
```

Comment:
It would be good test test failure at other stages to make sure the stats looks correct.

Reply:
Done. More tests are added to test failures in req/resp headers/body/trailer.